File: src/Algorithm/LeetCode/LeetCode_53_Maximum_Subarray/Solution_DP.java
package Algorithm.LeetCode.LeetCode_53_Maximum_Subarray;

class Solution_DP {
    // For SUBARRAY, you CANNOT skip the current element because the subarray has to be continuous.
    // You need to either include the previous subarray, or start it over
    public int maxSubArray(int[] nums) {
        int pre = 0;
        int max = nums[0];

        for (int n : nums) {
            pre = Math.max(n, pre + n);
            max = Math.max(pre, max);
        }

        return max;
    }
}

File: src/Algorithm/LeetCode/LeetCode_53_Maximum_Subarray/Solution_DP.py
from typing import List


# https://www.programmercarl.com/0053.最大子序和（动态规划）.html
class Solution_DP:
    def maxSubArray(self, nums: List[int]) -> int:
        result = 0
        length = len(nums)
        dp = [0] * length

        dp[0] = nums[0]
        result = dp[0]

        for i in range(1, length):
            dp[i] = max(dp[i - 1] + nums[i], nums[i])
            result = max(result, dp[i])

        return result
